### ISSUE TITLE
v3 question: how to complete with list instead of queryset?

### ISSUE THREAD

**Issue by dmosberger** [yourlabs/django-autocomplete-light#635]
Hopefully I'm missing something obvious, but it's not clear to me how to use a list of choices (rather than a queryset) for autocompletion.  For example, in v2, I could write:

```
class Foo_Autocomplete(al.AutocompleteListBase):
 choices = [ 'choice A', 'choice B', 'choice C' ]
 ...
```
How can I do the same with v3?

**Comment by jpic** [comment]
It's not demonstrated yet, but you could use a ChoiceField instead of a ModelChoiceField and in the view use reuse the mixins you want.

Are you really using a list or are you using another backend (ie. redis) or something ?

**Comment by dmosberger** [comment]
Yes, in the project I'd like to do that, we want to autocomplete, e.g., on the list of (electronic part) footprints.  Those footprints are generated for an EDA program (e.g., kicad), so it doesn't really make sense to represent them in a model.  Instead, we just construct a list of available footprints by walking a local directory containing them.

Any chance to see some sample code?  I'm not that familiar with the inner workings of dal3.

**Comment by jpic** [comment]
Yeah perhaps, will you be able to contribute documentation if you have example code ?

**Comment by dmosberger** [comment]
Sure, seems like a fair deal!

**Comment by jpic** [comment]
PR uploaded !

Note that I did a couple of minor changes in the package, if you don't want to do them then you can import ViewMixin from dal.views, and put the Select2 class directly in your form (it contains no code).

**Comment by jpic** [comment]
As mentioned in the PR, note that you don't even have to pass a url to the widget if you don't need a view ! It really depends, how much options do you want to propose ? If < 300 I'd say just go for :

```
class Select2(WidgetMixin, Select2WidgetMixin, forms.Select):
    """Select widget for regular choices and Select2."""


class TestForm(forms.ModelForm):
    test = forms.ChoiceField(
        choices=[
            ('windows', 'windows'),
            ('linux', 'linux'),
        ],
        required=False,
        widget=autocomplete.Select2(url='select2_list')
    )

    class Meta:
        model = TestModel
        fields = ('name', 'test')

```

Now feel free to go ahead and share some creativity to present that in nice documentation for the community with :heart: B)

**Comment by dmosberger** [comment]
Cool, let me play with this to make sure I understand it.

**Comment by dmosberger** [comment]
OK, I'm starting to understand it (I hope).  It took me a while to realize I needed to copy select2-4.0.2/dist to dal_select2/static/autocomplete_light/vendor/select2/.

Now I can do:

```
class Footprint_Autocomplete(autocomplete.ViewMixin, generic.View):
    choices = Footprints.get()

    def get(self, request, *args, **kwargs):
        assert_permission(self.request, perms.VIEW)
        result = choices
        if self.q:
            result = [ x for x in result if self.q in x ]
        return HttpResponse(json.dumps({
            'results': [dict(id=x, text=x) for x in result]
        }))

    @staticmethod
    def widget():
        return autocomplete.Select2(url='epic:footprint-dal',
                                    attrs={
                                        'data-placeholder': 'Footprint?'
                                    })
```

It would be nicer if there could be an autocomplete.Select2ListView which would take care of converting the list to a JSON HTTPResponse (and maybe the method should be called get_list() instead of just get()?).

Anyhow, it's getting late here, so I'll leave that for tomorrow.

Thanks for the help!

**Comment by jpic** [comment]
> OK, I'm starting to understand it (I hope). It took me a while to realize I needed to copy select2-4.0.2/dist to dal_select2/static/autocomplete_light/vendor/select2/.

Well, you should have cloned with --recursive or updated submodules. But don't worry you're not the first to be bitten by this, #638 includes a curative fix ^^

> It would be nicer if there could be an autocomplete.Select2ListView which would take care of converting the list to a JSON HTTPResponse (and maybe the method should be called get_list() instead of just get()?).

Agreed ! Feel free to open a PR with a base list view and refactor existing views accordingly ;)

Best ;)

**Comment by dmosberger** [comment]
How about something along these lines:
```
diff --git a/src/dal/autocomplete.py b/src/dal/autocomplete.py
index 2d916c9..3d59bca 100644
--- a/src/dal/autocomplete.py
+++ b/src/dal/autocomplete.py
@@ -21,7 +21,7 @@ from django.conf import settings as django_settings
 
 from .forms import FutureModelForm
 
-from .views import ViewMixin
+from .views import ViewMixin, Select2ListView
 
 from .widgets import (
     Select,
diff --git a/src/dal/views.py b/src/dal/views.py
index 13dcf34..ecede0c 100644
--- a/src/dal/views.py
+++ b/src/dal/views.py
@@ -6,7 +6,7 @@ from django import http
 from django.contrib.auth import get_permission_codename
 from django.core.exceptions import ImproperlyConfigured
 from django.utils import six
-from django.views.generic.list import BaseListView
+from django.views.generic.list import View, BaseListView
 
 
 class ViewMixin(object):
@@ -30,6 +30,36 @@ class ViewMixin(object):
         return super(ViewMixin, self).dispatch(request, *args, **kwargs)
 
 
+class Select2ListView(ViewMixin, View):
+    create_field = False
+
+    def get_list(self):
+        return []
+
+    def get(self, request, *args, **kwargs):
+        results = self.get_list()
+        create_option = []
+        if self.q:
+            results = [ x for x in results if self.q in x ]
+            if self.create_field:
+                create_option = [{
+                        'id': self.q,
+                        'text': 'Create "%s"' % self.q,
+                        'create_id': True
+                    }]
+        return http.HttpResponse(json.dumps({
+            'results': [dict(id=x, text=x) for x in results] + create_option
+        }))
+
+    def post(self, request):
+        text = request.POST.get('text', None)
+        if text is None:
+            return http.HttpResponseBadRequest()
+        return http.HttpResponse(json.dumps({
+            'id': text,
+            'text': text,
+        }))
+
 class BaseQuerySetView(ViewMixin, BaseListView):
     """Base view to get results from a QuerySet.

```

Obviously, the new code should probably be in dal_select2/views.py but other than that, hopefully it's fine.

With this, I can then do:
```
class Footprint_Autocomplete(autocomplete.Select2ListView):
    choices = Footprints.get()
    create_field = True

    def get_list(self):
        assert_permission(self.request, perms.VIEW)
        return self.choices

    @staticmethod
    def widget():
        return autocomplete.Select2(url='epic:footprint-dal',
                                    attrs={
                                        'data-placeholder': 'Footprint?'
                                    })
```

Here, "create_field" is simply a boolean, since no objects actually have to be created when adding a new item.  With create_field==False, select2 will say "No results" when an unknown query is typed.  With create_field=True, select2 will offer a choice called create "query".

**Comment by guettli** [comment]
Thank you jpic and dmosberger: I am interested in the more bare-bone access to autocomplete, too.

**Comment by mwhavens** [comment]
Thanks for this dmosberger. This feature addition is exactly what I need for a project I'm working on and I'm anxiously awaiting this improvement being adopted into the official release. Please let me know if there's anything I can do to help you.

**Comment by dmosberger** [comment]
James did most of the work, really!

Unfortunately, I got pulled into another project that's going to require my
full attention for a while.  I'll definitely get back to EPIC (and
therefore autocomplete), but it'll be a while.  I think James was mostly
looking for more help with testing the changes, so if that's something you
can help with, I'm sure he'd appreciate it.

  --david

On Tue, Apr 26, 2016 at 11:42 AM, mwhavens <notifications@github.com> wrote:

> Thanks for this dmosberger. This feature addition is exactly what I need
> for a project I'm working on and I'm anxiously awaiting this improvement
> being adopted into the official release. Please let me know if there's
> anything I can do to help you.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/yourlabs/django-autocomplete-light/issues/635#issuecomment-214824009>
>



-- 
Love Linux kernel hacking?  Know Python?  We are hiring!  Contact us at
jobs@egauge.net

**Comment by mwhavens** [comment]
Thanks for all the work on this James. Please let me know what testing I can help with.

**Comment by jpic** [comment]
Thanks but I can't take credit for all the contributions here <3

We just need the new code to be covered by tests, and executable with a command such as `tox -e base-py27-django19-sqlite` and we're good to go B)

**Comment by thecardcheat** [comment]
I am working on completing the list-autocompletion branch so avoid snatching without letting me know please ;)

Need it for a dependent project so ETA as I have time to figure out how the tests fit together and feel comfortable with the implementation - not too long I hope.

**Comment by thecardcheat** [comment]
In #702 

Since Select2 is not a widget that can work with TextField, using a ChoiceField is appropriate but requires you to override validate() if you want to allow values (by creating them) that are not in the choices provided when the instance is created. It seems we should have two custom fields to prevent developer confusion, one to accept a list and unpack it into a working iterable for ChoiceField since there is no reasonable way to handle text and values that do not match, and another that follows this same behavior but skips choice validation so text can be created.

The documentation should also be updated with this info as well as to clarify that if you do not use a common source for your Field and Select2ListView (unless of course you want to use the field that skips validation) you are asking for pain.

Thoughts?

**Comment by thecardcheat** [comment]
PR #738 is up. Let me know if that isn't how you saw it going.

**Comment by jpic** [comment]
I'd be very careful when creating values in validate(). This is something we tried in 3.0 but that caused problems when a user would submit a form that doesn't validate, but still creates a choice in the back end. The thing is, validate() is for validating a result without writing to the back end.

**Comment by thecardcheat** [comment]
Nothing should be getting created in validate(), the custom field that [defines a new validate()](https://github.com/yourlabs/django-autocomplete-light/pull/738/commits/2d103c508114b5c690053ce4f13b76202eb17942#diff-ab549cd76341438729203424d4021fd7R35) makes it a pass instead of a check that all choices are valid otherwise raising a ValidationError. I do notice now the call to super validate() in [OG ChoiceField](https://docs.djangoproject.com/en/1.10/_modules/django/forms/fields/#ChoiceField) which checks for empty values if self.required, so we need to add that to be consistent.

**Comment by jpic** [comment]
We've merged it ! #745
